Changes to support input sequence ID tracking

alexmsettle opened the pull request:
in the NVTX markers.  This feature adds additional information
to the NVTX marker string eg seq_ids=[101, 102, 103].  This indicates
the sequence id of the op which produced the input tensor based on its
position index in the array.  In the above example input tensor 0 was produced by 
the node with sequence id 101, input tensor 1 is from node 102, input tensor 2 is from 
node with sequence id 103. This is the same way the sizes array is
organized. If you know the sequence id of the node and the sequence ids
of the input edges, then you have enough information to construct the
network graph.

Fixes #66105

alexmsettle committed:
Changes to support input sequence ID tracking

in the NVTX markers.  This feature adds additional information
to the NVTX marker string eg seq_ids=[101, 102, 103].  This indicates
the sequence id of the op which produced the input tensor based on its
position index in the array.  This is the same way the sizes array is
organized. If you know the sequence id of the node and the sequence ids
of the input edges, then you have enough information to construct the
network graph.

alexmsettle committed:
Fixed build error

alexmsettle committed:
Fixed cuda device regression error

alexmsettle committed:
Fixed lint error

alexmsettle committed:
Fixed lint error in test

alexmsettle committed:
Removed print from new test

alexmsettle committed:
Removed comment in test - lint error

alexmsettle commented:
Hi - this is my first PR for pytorch.  What is the next step for getting this through the CI process?

robieta commented:
> Hi - this is my first PR for pytorch. What is the next step for getting this through the CI process?

Hey, thanks for the PR. I'm on vacation until the 7th, and then I'll review and help guide it in.

alexmsettle commented:
Thanks!

On Tue, Jan 4, 2022 at 6:04 PM Taylor Robie ***@***.***>
wrote:

> Hi - this is my first PR for pytorch. What is the next step for getting
> this through the CI process?
>
> Hey, thanks for the PR. I'm on vacation until the 7th, and then I'll
> review and help guide it in.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/70264#issuecomment-1005318387>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI5QN2TLKCC5KPX7VMC6ME3UUORJNANCNFSM5KRGFM6Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

alexmsettle committed:
Fixed merge conflicts for #70264

Merge remote-tracking branch 'origin/master' into alex_input_seq_ids

alexmsettle committed:
Updated string type for #70264

alexmsettle commented:
I just resolved a merge conflict that was introduced after I started this PR.  It looks like the kineto sw has been changing recently and it is causing conflicts with my changes.

robieta commented:
We just wrapped up performance reviews, so I'll finally have time to review this. Thanks for your patience.

alexmsettle commented:
Thanks!

Any idea how to restart a the test? I saw that one of the CI jobs was
cancelled.

-Alex

On Tue, Jan 18, 2022 at 8:39 AM Taylor Robie ***@***.***>
wrote:

> We just wrapped up performance reviews, so I'll finally have time to
> review this. Thanks for your patience.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/70264#issuecomment-1015593552>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI5QN2XFROTPXQDW3ONBMFDUWWJVBANCNFSM5KRGFM6Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

robieta commented:
> Thanks! Any idea how to restart a the test? I saw that one of the CI jobs was cancelled.
> […](#)
> -Alex
> On Tue, Jan 18, 2022 at 8:39 AM Taylor Robie ***@***.***> wrote: We just wrapped up performance reviews, so I'll finally have time to review this. Thanks for your patience. — Reply to this email directly, view it on GitHub <[#70264 (comment)](https://github.com/pytorch/pytorch/pull/70264#issuecomment-1015593552)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AI5QN2XFROTPXQDW3ONBMFDUWWJVBANCNFSM5KRGFM6Q> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you authored the thread.Message ID: ***@***.***>

I went ahead and just manually restarted it.

alexmsettle commented:
Thanks for restarting the pipeline, looks like it worked this time.

robieta commented:
At a high level, this seems awesome! So much so that I also want this for the Kineto profiler. (Though I'm perfectly happy to have this PR just add it for NVTX and then I can steal all the utility functions later.)

I do have some concerns about using Autograd Node as a proxy for Tensor identity. I understand why it's appealing since Tensors don't have a good way of assigning uuid, but I worry about losing information. For one, this won't work in inference mode (which, ironically, is when a lot of more intrusive optimizations like buffer reuse become available). It also (I think) will alias Tensors that are produced from multi-output ops like `split` unless we also record the output index. (Though `[sequenceID, TensorImpl*]` should be unambiguous.) And lastly, because of views and version bumps in autograd some analyses (like memory planning or fusion) won't be possible unless we also trace `StorageImpl* / storage_offset 
/ strides / version_number`. WDYT?

alexmsettle commented:
Thanks for the feedback @robieta !

I think your concerns about using the autograd Node are valid.  When I added this feature I was specifically working on performance analysis of training, so not working on inference wasn't really a concern at the time.   Also the Autograd Node was the only data structure I could find that maintained information about the producer of a given tensor.  Leveraging the JIT graph is another option I looked into, but it was difficult to convert the entire network to torch script for all the nets I was investigating.  I would like for this to work for inference, I'm actually starting to analyze inference performance now, so it is an important issue for me.   Do you have any suggestions for alternative data structures for maintaining the tensors' producer information?

For training the autograd Node approach works reasonably well.  I have run into some issues that would be good to resolve.   Would it be possible to move forward with this PR, then work on a better solution to recording the tensors' producer?  Most of the code supporting the profiling would stay in-tact.  The LOC for the autograd node approach is small, so replacing it with a more robust solution shouldn't be too much effort once there's s a good solution in place.

robieta commented:
FYI: https://github.com/pytorch/pytorch/pull/71135 will hopefully make this easier since you'll be able to hack on just the NVTX observer without having to deal with the (much more complex) Kineto code path.

I agree that you don't want to rely on the JIT graph. Probably the best way to track identity would be to map `TensorImpl*` to an index, and have the observer record outputs as well. You can use `reportMemoryUsage` to solve the ABA problem. The one downside compared to autograd is that you don't get info about Tensors which are created when profiling is disabled, but that shouldn't be too big a deal.

I'm a little bit wary of baking things into the API that we expect to change immediately; however the NVTX string part of the profiler is kind of wild west anyway, so if you feel strongly about getting a v1.0 with sequence IDs let me know and we can proceed on that basis.

alexmsettle committed:
Fixed merge conflicts

Merge remote-tracking branch 'origin/master' into alex_input_seq_ids

alexmsettle committed:
Updated the way input nodes are tracked for #72064

alexmsettle commented:
Hi @robieta I just pushed an update to how I keep track of the input op ids.  It took me a little while to figure out how to get the address of a TensorImpl.  I added logic to aten/src/ATen/record_function.[cpp,h] that maps the TensorImpl address of the producer op to the unique <record function, output number> pair.  The record functions already have a unique id associated with them state_->handle() so I'm using that for the op id.  I also had to enable 'needsOutputs' because there is conditional logic which guards at::RecordFunction::setOutputs().  One challenge I found with tracking the outputs is that it looks like the actual tensors get freed automatically by a reference counter before the profiler logic can inspect them.  I added some extra variables in RecordFunction to keep track of the output tensor shapes and the mapping I described above, so the profiler isn't dependent on the actual output tensor.

These changes are still WIP.  I need to connect this new logic to the profilers and remove the old logic that uses autograd Node.  Let me know what you think.

alexmsettle committed:
Fixed conflicts

Merge remote-tracking branch 'origin/master' into alex_input_seq_ids

alexmsettle committed:
Enabled the new input op_id tracking software

in both kineto and legacy profilers

alexmsettle committed:
Connected new c++ variables op_id and input_op_ids (#70264)

to python front end.

alexmsettle commented:
Hi @robieta these changes are working now.  I just need to remove the code from the original implementation before it is ready to commit.  Can you take a look at the changes I made?

alexmsettle committed:
Removed references to seq_ids.  Replaced by input_op_ids

alexmsettle committed:
Changed type of input op_ids from uint to signed int

alexmsettle committed:
Cleaned up debug code

alexmsettle committed:
Fixed input shape tracking error for cat ops.

alexmsettle committed:
Fixed regression error segfault in profiler_kineto

alexmsettle committed:
Fixed CI regression errors

Lint error in util.cpp
Test error in test_profiler.py.  Modified the test case to account
for a bug fix I made to tracking the input tensor shapes for concat
ops.

alexmsettle committed:
Fixed lint error, removed some dead code

robieta commented:
I don't think this logic should live in RecordFunction, as we want to keep it as lean as possible. (Both from an API and implementation standpoint.) Am I understanding correctly that this is just because of the issue with outputs being freed? (If so, we should just fix the refcount issue.) I also thought this was just for the NVTX observer; we don't want to make changes to the legacy profiler, and this will conflict with changes that are coming on the kineto profiler. I also don't think we should have a global `producer_tensor_map` object; rather any tracking should be scoped to a profiler TLS. I'm still going through the code to review more implementation detail level questions.

alexmsettle commented:
I'll move the logic from record_record function to the nvtx profiler.  Right the reason I ended up putting this logic in record_function is that the output tensor was getting freed before the profiler callback functions could observe it.

I can also remove the changes I made to kineto profiler.  I only need this feature to be available in nvtx observer.  The first time I implemented this, there was a dependency on a data structure in kineto profiler, so I just updated both profilers.

I'll change the scope of producer_tensor_map too.

alexmsettle committed:
Removed input_op_id tracking code from

profiler_kineto and profiler_legacy.  Now only the
nvxt observer will track the input_op_ids.

alexmsettle committed:
Moved input_op_id tracking sw from record_function

to the nvtx_observer.  This keeps all the tracking sw
local to the thread TLS.  Record function already provides
enough information in the input and output tensors in order
to enable tracking of which ops produced the input tensors.

alexmsettle committed:
Removed a test for input_op_ids that is no

longer valid.

alexmsettle commented:
Hi @robieta I just made the changes you recommended.  I moved the producer tensor logic out of record_function and into the nvtx observer.  The producer_tensor_map now has TLS scope.  I also removed the code changes I had made to legacy profiler and kineto profiler.  It turns out that the problem I had with observing the output tensors was not related to memory being freed unexpectedly.  Instead the output tensor information was included in the end callbacks.  I was only checking the start callback.  I just added some logic to nvtx observer that does the output tensor tracking during the end callback.

alexmsettle committed:
Fixed lint error (#70264)

alexmsettle committed:
Fixed CI errors

alexmsettle committed:
Fixed CI errors

alexmsettle committed:
Fixed CI test

alexmsettle commented:
hi @robieta  - I'm just wrapping up debugging some failing tests.  Could you please review the changes?  I moved most of the op id tracking sw into nvxt_observer.

alexmsettle committed:
Merge branch 'pytorch:master' into alex_input_seq_ids

alexmsettle committed:
Merge branch 'pytorch:master' into alex_input_seq_ids

alexmsettle commented:
Hi @robieta  - any idea what could be causing the 2 CI build errors from Windows configs?

alexmsettle committed:
Merge branch 'pytorch:master' into alex_input_seq_ids

robieta commented on torch/csrc/profiler/util.cpp:
```diff
@@ -115,20 +106,47 @@ std::string stacksToStr(
 std::vector<std::vector<int64_t>> inputSizes(const at::RecordFunction& fn) {
   std::vector<std::vector<int64_t>> sizes;
   sizes.reserve(fn.inputs().size());
+  int iter = 0;
```
unused?

alexmsettle replied:
removed

robieta commented on torch/csrc/profiler/util.cpp:
```diff
@@ -115,20 +106,47 @@ std::string stacksToStr(
-    const at::Tensor& tensor = input.toTensor();
-    if (tensor.defined()) {
-      sizes.push_back(input.toTensor().sizes().vec());
+      if (input.isList()) {
```
nit: At this point is should probably be `if (input.isTensor) { ... } else if (input.isList()) { ... } else {}`
Also, can we add a `flatten_lists` arg so we don't change the behavior of other profilers? Otherwise we could break users who previously ignored lists, and now have no way to reconstruct the old argument setup.

alexmsettle replied:
I simplified the if clause like you proposed.
Also I added a flag to inputSizes  `flatten_list_enabled`.   It is false by default.  It is set to true by nvtx_observer.  When using nvtx_observer users can't directly query the profiler, so there shouldn't be existing code that is testing the shape string.  In the timeline view I need to see the flattened list in order to know the tensor shapes that were passed to the concat op.

robieta reviewed (changes requested):
This all seems reasonable. I have some nits, but nothing that requires any monumental change. Thanks!

chaekit commented on torch/csrc/profiler/util.cpp:
```diff
@@ -32,32 +35,20 @@ std::string getNvtxStr(
       str = name;
 #endif
     }
+    std::stringstream s;
```
we avoid `stringstream` when there is no input as it adds significant overhead. This will regress our our no-input collection path

alexmsettle replied:
removed stringstream, used fmt::format() w/ std::string instead

albanD commented on test/test_autograd.py:
```diff
@@ -8658,6 +8658,33 @@ def test_profiler_emit_nvtx(self, device):
 
+    @skipCUDAIfRocm
+    @onlyCUDA
+    def test_custom_module_input_op_ids(self, device):
```
Why isn't this in test_profiler?

alexmsettle replied:
Moved to test_profiler.py

albanD commented on torch/autograd/profiler_util.py:
```diff
@@ -495,7 +495,8 @@ def __repr__(self):
             '<FunctionEvent id={} name={} device_type={} node_id={} cpu_time={} start_us={} end_us={} '
             'cpu_children={} cuda_time={} name={} thread={} input_shapes={} '
-            'cpu_memory_usage={} cuda_memory_usage={} is_async={} is_remote={} seq_nr={} is_legacy={}>'.format(
+            'cpu_memory_usage={} cuda_memory_usage={} '
```
Is this changed?

alexmsettle replied:
Good catch, I had made a change then removed it.  I'm reverting back to the original, so this file will be unchanged.

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -705,6 +704,12 @@ PyObject *THPFunction_apply(PyObject *cls, PyObject *inputs)
   ctx->is_variable_input = std::move(input_info.is_variable_input);
 
+  // Call record function after all the inputs have been decoded.
+  RECORD_FUNCTION(
```
Early exits just above correspond to the creation of the backward object. So user code ran yet.
So I guess yes?

alexmsettle replied:
Not sure if I understand the comments.  Does there need to be a RECORD_FUNCTION before ctx is available?  I want to be able to see the input shapes for user defined ops.

robieta replied:
I think this means that you need to make a `RECORD_FUNCTION` without inputs for the early exit cases.

alexmsettle replied:
I moved RECORD_FUNCTION before the early exits.  I also moved the input processing above.  The inputs don't depend on context or backward state so it is safe to do that.

alexmsettle commented on torch/csrc/profiler/api.h:
```diff
@@ -80,6 +80,14 @@ struct TORCH_API ProfilerThreadLocalStateBase
     return config_;
   }
 
+  const std::unordered_map<at::TensorImpl*, std::pair<at::RecordFunctionHandle, int>>& producerTensorMap() const {
```
Moved to nvtx_observer.cpp

alexmsettle commented on torch/csrc/profiler/api.h:
```diff
@@ -80,6 +80,14 @@ struct TORCH_API ProfilerThreadLocalStateBase
+    return producer_tensor_map_;
+  }
+
+  void setProducerTensorMap(at::TensorImpl *tensor, int op_id, int output_nr){
```
made the type for op_id at::RecordFunctionHandle

alexmsettle replied:
and updated the functions downstream of it to also use at::RecordFunctionHandle.  It looks like the counting of handle starts at 1 so the value 0 can be used to mean invalid.

alexmsettle commented on torch/csrc/profiler/api.h:
```diff
@@ -105,6 +113,9 @@ struct TORCH_API ProfilerThreadLocalStateBase
   at::CallbackHandle handle_ = 0;
+  // Maps the address of an output Tensor to a unique op id and output index of the tensor
+  // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
+  std::unordered_map<at::TensorImpl*, std::pair<at::RecordFunctionHandle, int>> producer_tensor_map_;
```
Made the key void*

alexmsettle commented on torch/csrc/profiler/nvtx_observer.cpp:
```diff
@@ -32,15 +32,98 @@ struct NVTXThreadLocalState : ProfilerThreadLocalStateBase {
   }
 };
 
+std::pair<int, int> getOpIdFromInput(const c10::IValue& input_item) {
```
changed arg to const Tensor&

alexmsettle commented on torch/csrc/profiler/nvtx_observer.cpp:
```diff
@@ -32,15 +32,98 @@ struct NVTXThreadLocalState : ProfilerThreadLocalStateBase {
+std::pair<int, int> getOpIdFromInput(const c10::IValue& input_item) {
+  std::pair<int, int> producer_op_pair(-1, -1);
+  const at::Tensor& tensor = input_item.toTensor();
+  auto state_ptr = NVTXThreadLocalState::getTLS();
```
I can move getOpIdFromInput to NVTXThreadLocalState, but I don't think it will work for updateOutputTensorTracker.  The reason is updateOutputTensorTracker is called from the 2nd arg to RecordFunctionCallback().  This arg is a function pointer that only takes RecordFunction and Ctx as args.  I would need to also pass in state_ptr to call updateOutputTensorTracker, but that would mean changing the signature to RecordFunctionCallback().

alexmsettle commented on torch/csrc/profiler/nvtx_observer.cpp:
```diff
@@ -32,15 +32,98 @@ struct NVTXThreadLocalState : ProfilerThreadLocalStateBase {
+  std::pair<int, int> undefined_op_pair(-1,-1);
+  std::vector<std::pair<int, int>> input_producer_ops_;
+  input_producer_ops_.reserve(num_inputs);
+  int idx = 0;
```
removed it

alexmsettle commented on torch/csrc/profiler/nvtx_observer.cpp:
```diff
@@ -32,15 +32,98 @@ struct NVTXThreadLocalState : ProfilerThreadLocalStateBase {
+  input_producer_ops_.reserve(num_inputs);
+  int idx = 0;
+  for (const c10::IValue& input_item : fn.inputs()) {
+    bool tensor_valid = false;
```
removed

alexmsettle commented on torch/csrc/profiler/nvtx_observer.cpp:
```diff
@@ -32,15 +32,98 @@ struct NVTXThreadLocalState : ProfilerThreadLocalStateBase {
+        std::vector<std::pair<int, int>> tmp_op_ids = flattenOpIdList(input_item.toList(), std::string(fn.name()));
+        // Extend the current sizes array by the array returned from input sizes
+        if (!tmp_op_ids.empty()) {
+          input_producer_ops_.insert(input_producer_ops_.end(), tmp_op_ids.begin(), tmp_op_ids.end());
```
changed the type from std:vector to std::list and used list::slice to concat the lists.

alexmsettle commented on torch/csrc/profiler/nvtx_observer.cpp:
```diff
@@ -65,10 +148,13 @@ void pushNVTXCallbacks(
-          [](const at::RecordFunction&, at::ObserverContext*) {
+          [](const at::RecordFunction& fn, at::ObserverContext* ctx) {
             torch::profiler::impl::cudaStubs()->nvtxRangePop();
+            updateOutputTensorTracker(fn);
```
I don't think you can pass report_input_shapes to the lambda function.  If there are no output tensors in the RecordFunction, then updateOutputTensorTracker will just return, due to the output tensor list being empty.

alexmsettle commented on torch/csrc/profiler/util.cpp:
```diff
@@ -115,20 +106,47 @@ std::string stacksToStr(
 
+std::vector<std::vector<int64_t>> flattenList(c10::List<c10::IValue> list, std::string fn_name) {
+  std::vector<std::vector<int64_t>> tensor_dims;
+  int idx = 0;
```
removed

alexmsettle commented on torch/csrc/profiler/nvtx_observer.cpp:
```diff
@@ -32,15 +32,98 @@ struct NVTXThreadLocalState : ProfilerThreadLocalStateBase {
-                                : std::vector<std::vector<int64_t>>())
+                                : std::vector<std::vector<int64_t>>(),
+            int(fn.handle()),
+            report_input_shapes ? input_op_ids
```
correct.   The input shapes and input op ids go together.  I think you would always want to see input shapes if input_op_ids are being collected.

alexmsettle committed:
Code review updates (#70264)

alexmsettle committed:
Fixed build error for std::list

Fixed issue in nvtx observer so it only expands lists
when the items are tensors.

alexmsettle committed:
Merge branch 'pytorch:master' into alex_input_seq_ids

alexmsettle commented:
Hi @robieta - all my code review edits are in and the CI passed.  Can you please review?

robieta commented:
> Hi @robieta - all my code review edits are in and the CI passed. Can you please review?

Sure thing. I'm slightly swamped right now but I'll get to it as soon as I can.

alexmsettle commented:
Hi @robieta were you able to look at my code review edits yet?

alexmsettle commented:
ping @robieta  - can you please review these changes?

robieta reviewed (commented):
LGTM modulo the one comment on `python_function.cpp`. I'm going to import and start running the internal tests.

alexmsettle committed:
Moved RECORD_FUNCTION and input decode above

the early exits in THPFunction_apply().  Feedback
from code review.

alexmsettle committed:
Merge remote-tracking branch 'origin/alex_input_seq_ids' into alex_input_seq_ids

Remote alex_input_seq_ids was already merged with master, this merge makes
local branch match the remote.

alexmsettle committed:
Merge branch 'pytorch:master' into alex_input_seq_ids

alexmsettle committed:
Merge branch 'pytorch:master' into alex_input_seq_ids

alexmsettle commented:
Hi @robieta I fixed the RECORD_FUNCTION issue in python_function.cpp.  CI passed, so it should be good to go now.

robieta commented on torch/csrc/profiler/util.cpp:
```diff
@@ -134,23 +147,37 @@ std::vector<std::vector<int64_t>> inputSizes(const at::RecordFunction& fn) {
+  std::string str("[");
+  int idx = 0;
+
+  for (const auto op_id_info_pair : input_op_ids) {
```
Can you change this to `const auto& op_id_info_pair`? One of the internal builds is super pedantic and refuses to build if there is a copy.

alexmsettle replied:
I just pushed the change

alexmsettle committed:
Code review fix

alexmsettle committed:
Merge branch 'pytorch:master' into alex_input_seq_ids

alexmsettle commented:
@robieta Any idea what this error is?

Using default tag: latest
latest: Pulling from tool/alpine
Digest: sha256:def822f9851ca422481ec6fee59a9966f12b[35](https://github.com/pytorch/pytorch/runs/5565030763?check_suite_focus=true#step:6:35)1c62ccb9aca841526ffaa9f748
Status: Image is up to date for 308535[38](https://github.com/pytorch/pytorch/runs/5565030763?check_suite_focus=true#step:6:38)5114.dkr.ecr.us-east-1.amazonaws.com/tool/alpine:latest
308535385114.dkr.ecr.us-east-1.amazonaws.com/tool/alpine:latest
chown: ./docs/cpp/build/doctrees/environment.pickle: No such file or directory
Error: Process completed with exit code 1.

alexmsettle commented:
Any update on the internal testing?

robieta commented:
> Any update on the internal testing?

I was out the second half of last week. It's failing some tests, but unclear if it is unrelated flakes. I fired off a new batch.

alexmsettle commented:
Ok thanks for the update!

alexmsettle commented:
Any updates on the testing status?

alexmsettle commented:
ping @robieta

robieta commented:
I'm trying to get it through CI. Because so many projects have PyTorch as a dependency it is an involved process to separate flakes from legitimate breakages. Never fear, I am actively landing this PR.

alexmsettle commented:
> I'm trying to get it through CI. Because so many projects have PyTorch as a dependency it is an involved process to separate flakes from legitimate breakages. Never fear, I am actively landing this PR.

Much appreciated!

alexmsettle commented:
FYI @robieta  here's a clip from a network graph that I generated based on this PR.  This is from the network efficientnet.  I combined the op_ids with the module names from the nn.module object associated with each aten:: op.  It makes it really easy to understand the network architecture, you don't even really need to look at the network source code.

![image](https://user-images.githubusercontent.com/37422826/161144551-1a03fd02-465d-4f3a-9e94-69de9b203ae4.png)

alexmsettle commented:
> @alexmsettle Testing is in good enough shape (I'm quite confident that all remaining failures are unrelated) that I just started to land.

Awesome!  Thanks for driving this to completion @robieta
